File: CarrelApp/Core/Network/Models/Paper.swift
import Foundation

struct Paper: Codable, Identifiable, Equatable {
    let id: String
    let title: String?
    let authors: String?
    let pdfUrl: String?
    let thumbnailUrl: String?
    let status: PaperStatus
    let isPublic: Bool
    let shareSlug: String?
    let repositoryId: String?
    let trackedFileId: String?
    let lastSyncedAt: Date?
    let createdAt: Date
    let updatedAt: Date

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case title, authors, pdfUrl, thumbnailUrl, status
        case isPublic, shareSlug, repositoryId, trackedFileId
        case lastSyncedAt, createdAt, updatedAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        title = try container.decodeIfPresent(String.self, forKey: .title)
        authors = try container.decodeIfPresent(String.self, forKey: .authors)
        pdfUrl = try container.decodeIfPresent(String.self, forKey: .pdfUrl)
        thumbnailUrl = try container.decodeIfPresent(String.self, forKey: .thumbnailUrl)
        status = try container.decodeIfPresent(PaperStatus.self, forKey: .status) ?? .unknown
        isPublic = try container.decodeIfPresent(Bool.self, forKey: .isPublic) ?? false
        shareSlug = try container.decodeIfPresent(String.self, forKey: .shareSlug)
        repositoryId = try container.decodeIfPresent(String.self, forKey: .repositoryId)
        trackedFileId = try container.decodeIfPresent(String.self, forKey: .trackedFileId)

        // Parse timestamps (Convex sends milliseconds)
        if let lastSyncedMs = try container.decodeIfPresent(Double.self, forKey: .lastSyncedAt) {
            lastSyncedAt = Date(timeIntervalSince1970: lastSyncedMs / 1000)
        } else {
            lastSyncedAt = nil
        }

        if let createdMs = try container.decodeIfPresent(Double.self, forKey: .createdAt) {
            createdAt = Date(timeIntervalSince1970: createdMs / 1000)
        } else {
            createdAt = Date()
        }

        if let updatedMs = try container.decodeIfPresent(Double.self, forKey: .updatedAt) {
            updatedAt = Date(timeIntervalSince1970: updatedMs / 1000)
        } else {
            updatedAt = Date()
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encodeIfPresent(title, forKey: .title)
        try container.encodeIfPresent(authors, forKey: .authors)
        try container.encodeIfPresent(pdfUrl, forKey: .pdfUrl)
        try container.encodeIfPresent(thumbnailUrl, forKey: .thumbnailUrl)
        try container.encode(status, forKey: .status)
        try container.encode(isPublic, forKey: .isPublic)
        try container.encodeIfPresent(shareSlug, forKey: .shareSlug)
        try container.encodeIfPresent(repositoryId, forKey: .repositoryId)
        try container.encodeIfPresent(trackedFileId, forKey: .trackedFileId)
    }
}

enum PaperStatus: String, Codable {
    case synced
    case pending
    case building
    case error
    case unknown

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        let value = try container.decode(String.self)
        self = PaperStatus(rawValue: value) ?? .unknown
    }
}

struct PapersResponse: Codable {
    let papers: [Paper]

    init(from decoder: Decoder) throws {
        // The API returns an array directly
        let container = try decoder.singleValueContainer()
        papers = try container.decode([Paper].self)
    }
}

File: CarrelAndroid/app/src/main/java/com/carrel/app/core/network/models/Paper.kt
package com.carrel.app.core.network.models

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Paper(
    @SerialName("_id") val id: String,
    val title: String? = null,
    val authors: String? = null,
    val pdfUrl: String? = null,
    val thumbnailUrl: String? = null,
    val isUpToDate: Boolean? = null,
    val buildStatus: String? = null,
    val compilationProgress: String? = null,
    val lastSyncError: String? = null,
    val isPublic: Boolean = false,
    val shareSlug: String? = null,
    val repositoryId: String? = null,
    val trackedFileId: String? = null,
    val pdfSourceType: String? = null,
    // Convex returns timestamps as doubles (milliseconds with decimal precision)
    val lastAffectedCommitTime: Double? = null,
    val lastAffectedCommitAuthor: String? = null,
    val lastSyncedAt: Double? = null,
    val createdAt: Double? = null,
    val updatedAt: Double? = null
) {
    // Derive status from isUpToDate and buildStatus
    val status: PaperStatus
        get() = when {
            buildStatus == "building" -> PaperStatus.BUILDING
            buildStatus == "error" -> PaperStatus.ERROR
            buildStatus == "pending" -> PaperStatus.PENDING
            isUpToDate == true -> PaperStatus.SYNCED
            isUpToDate == false -> PaperStatus.PENDING
            else -> PaperStatus.UNKNOWN
        }
}

@Serializable
enum class PaperStatus {
    @SerialName("synced") SYNCED,
    @SerialName("pending") PENDING,
    @SerialName("building") BUILDING,
    @SerialName("error") ERROR,
    @SerialName("unknown") UNKNOWN;

    companion object {
        fun fromString(value: String): PaperStatus {
            return entries.find { it.name.equals(value, ignoreCase = true) } ?: UNKNOWN
        }
    }
}
